PPTP-752 Integrate with NRS

PauloPortugal opened the pull request:
NRS stands for Non Repudiation Service

Also involved is to separate successful fields from the failures, this provides us with a cleaner interface

RyanKellyHMRC commented on test/uk/gov/hmrc/plasticpackagingtaxregistration/connectors/NonRepudiationConnectorISpec.scala:
```diff
@@ -0,0 +1,97 @@
+
+    }
+  }
+
```
Do we need tests for unsuccessful responses?

RyanKellyHMRC commented on test/uk/gov/hmrc/plasticpackagingtaxregistration/controllers/SubscriptionControllerSpec.scala:
```diff
@@ -89,7 +93,8 @@ class SubscriptionControllerSpec
+                                       withMetaDataRequest(MetaData(true, true)),
+                                       withUserHeaders(Map("testHeaderKey" -> "testHeaderValue"))
     )
 
```
I think we could do with some tests to prove that the NRS is called when a PPT submission is successful.

RyanKellyHMRC commented on test/uk/gov/hmrc/plasticpackagingtaxregistration/services/nrs/NonRepudiationServiceSpec.scala:
```diff
@@ -0,0 +1,160 @@
+                                                    searchKeys =
+                                                      Map("pptReference" -> testPPTReference)
+      )
+
```
Could some of this test data be shared across tests?

RyanKellyHMRC commented on test/uk/gov/hmrc/plasticpackagingtaxregistration/services/nrs/NonRepudiationServiceSpec.scala:
```diff
@@ -0,0 +1,160 @@
+
+      await(res) mustBe NonRepudiationSubmissionAccepted(testSubmissionId)
+    }
+    "audit when the non repudiation call fails" in {
```
I was expecting this test to verify that an audit was recorded because the NRS call failed. It appears to verify that an exception is thrown by the `NonRepudiationService` when the call fails. Is this auditing functionality to be done as part of a future ticket?

RyanKellyHMRC reviewed (changes requested):
I like your refactoring to use specific success and failure types in the `SubscriptionsConnector` although I don't see (and this is probably my lack of understanding of Scala and/or the Play Framework) how the failure type is ever returned to clients.

Elsewhere I think I've identified some opportunity for small improvements.

RyanKellyHMRC commented on app/uk/gov/hmrc/plasticpackagingtaxregistration/services/nrs/NonRepudiationService.scala:
```diff
@@ -0,0 +1,148 @@
+      }.recoverWith {
+        case exception: HttpException => Future.failed(exception)
+      }
+    } yield nonRepudiationSubmissionResponse
```
I wonder whether this ticket represents this : https://jira.tools.tax.service.gov.uk/browse/PPTP-868?

PauloPortugal replied:
Not part of this ticket part of PPTP-868 Retry mechanism. 
That requirement should have been on the other ticket.

PauloPortugal commented on app/uk/gov/hmrc/plasticpackagingtaxregistration/services/nrs/NonRepudiationService.scala:
```diff
@@ -0,0 +1,148 @@
+      userAuthToken = hc.authorization match {
+        case Some(Authorization(authToken)) => authToken
+        case _                              => throw new InternalServerException("No auth token available for NRS")
+      }
```
This has been addressed.

PauloPortugal commented:
> I was expecting this test to verify that an audit was recorded because the NRS call failed. It appears to verify that an exception is thrown by the NonRepudiationService when the call fails. Is this auditing functionality to be done as part of a future ticket?

Good spot. This will be the future test, once we move the TxM/CiP auditing from the FE to the BE.

> Could some of this test data be shared across tests?
Refactored the test code to reuse as much as possible test data. Used traits for it.

> I think we could do with some tests to prove that the NRS is called when a PPT submission is successful.
Done. Mocked the NRS Service from the NRS Controller.

> I know you've not changed things here but why do we return OK(200) when the PPT subscription has failed?
That was a bug introduced by me. The failures will be available on EIS/IF 4xx or 5xx responses and at the moment we are not displaying them to the user or doing anything with it.
Removed that case.

RyanKellyHMRC commented on test/uk/gov/hmrc/plasticpackagingtaxregistration/controllers/SubscriptionControllerSpec.scala:
```diff
@@ -89,48 +88,32 @@ class SubscriptionControllerSpec
-
-        response.failures.get.isEmpty mustBe false
-        verifyNoInteractions(mockRepository)
+        verify(mockNonRepudiationService).submitNonRepudiation(any(), any(), any(), any())(any())
```
Is it worth checking that the parameters passed to the `NonRepudiationService` are as expected (rather than using `any()`) or do you think this makes the test overly brittle?

RyanKellyHMRC commented on test/uk/gov/hmrc/plasticpackagingtaxregistration/controllers/SubscriptionControllerSpec.scala:
```diff
@@ -89,48 +88,32 @@ class SubscriptionControllerSpec
     )
 
     "return 200 and delete registration" when {
       "request is valid" in {
```
I think we should update these test names so that they better reflect what we are testing and the associated setup (test fixture). Does `Return 200, make NRS submission and delete temporary registration when PPT registration submission is successful` work better?

I think it would be even better to split the testing of `NRS submission` and `registration deletion` into discrete tests.

RyanKellyHMRC reviewed (approved):
Thanks for making the suggested changes @PauloPortugal. I made a couple of additional suggestions but please don't allow these to hold up merging if you don't feel strongly that these improve things sufficiently.

PauloPortugal commented:
>Is it worth checking that the parameters passed to the NonRepudiationService are as expected (rather than using any()) or do you think this makes the test overly brittle?

I have addressed that. good point. Done!

PauloPortugal commented:
> I think it would be even better to split the testing of NRS submission and registration deletion into discrete tests.

This is the only outstanding debt out of this ticket. 
Probably worth discussing at some point.
